File: paykit-mobile/android-demo/app/src/androidTest/java/com/paykit/demo/TestHelpers.kt
// TestHelpers.kt
// Test Helpers for PaykitDemo E2E Tests
//
// Provides common utilities, mock services, and test fixtures for E2E testing.

package com.paykit.demo

import java.util.UUID
import java.util.Date

// MARK: - Test Configuration

/**
 * Configuration for E2E tests
 */
object TestConfig {
    /** Test timeout for async operations (ms) */
    const val DEFAULT_TIMEOUT_MS = 30000L
    
    /** Short timeout for quick operations (ms) */
    const val SHORT_TIMEOUT_MS = 5000L
    
    /** Generate a random available port */
    fun randomPort(): Int = (10000..60000).random()
    
    /** Test user identifiers */
    fun testUserAlice() = "alice_test_${UUID.randomUUID().toString().take(8)}"
    fun testUserBob() = "bob_test_${UUID.randomUUID().toString().take(8)}"
}

// MARK: - Mock Key Manager

/**
 * Mock key manager for testing
 */
class MockKeyManager {
    private val identities = mutableMapOf<String, MockIdentity>()
    private var currentIdentity: String? = null
    
    data class MockIdentity(
        val nickname: String,
        val publicKeyZ32: String = "z32_${UUID.randomUUID().toString().replace("-", "").take(52)}",
        val secretKey: ByteArray = ByteArray(32) { (0..255).random().toByte() }
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as MockIdentity
            return nickname == other.nickname && publicKeyZ32 == other.publicKeyZ32
        }
        
        override fun hashCode(): Int {
            var result = nickname.hashCode()
            result = 31 * result + publicKeyZ32.hashCode()
            return result
        }
    }
    
    fun createIdentity(nickname: String): MockIdentity {
        val identity = MockIdentity(nickname = nickname)
        identities[nickname] = identity
        return identity
    }
    
    fun setCurrentIdentity(nickname: String): Boolean {
        if (!identities.containsKey(nickname)) return false
        currentIdentity = nickname
        return true
    }
    
    fun getCurrentIdentity(): MockIdentity? {
        return currentIdentity?.let { identities[it] }
    }
    
    fun getPublicKey(): String? {
        return getCurrentIdentity()?.publicKeyZ32
    }
}

// MARK: - Mock Receipt Store

/**
 * Mock receipt store for testing
 */
class MockReceiptStore {
    private val receipts = mutableMapOf<String, MockReceipt>()
    
    data class MockReceipt(
        val id: String,
        val payerPubkey: String,
        val payeePubkey: String,
        val amountSats: Long,
        val createdAt: Date = Date(),
        val status: String
    )
    
    fun storeReceipt(receipt: MockReceipt) {
        receipts[receipt.id] = receipt
    }
    
    fun getReceipt(id: String): MockReceipt? {
        return receipts[id]
    }
    
    fun getAllReceipts(): List<MockReceipt> {
        return receipts.values.toList()
    }
    
    fun clear() {
        receipts.clear()
    }
}

// MARK: - Mock Directory Service

/**
 * Mock directory service for testing endpoint discovery
 */
class MockDirectoryService {
    private val endpoints = mutableMapOf<String, MockEndpoint>()
    
    data class MockEndpoint(
        val recipientPubkey: String,
        val host: String,
        val port: Int,
        val serverNoisePubkey: String,
        val metadata: String? = null
    )
    
    fun publishEndpoint(
        pubkey: String,
        host: String,
        port: Int,
        noisePubkey: String,
        metadata: String? = null
    ) {
        endpoints[pubkey] = MockEndpoint(
            recipientPubkey = pubkey,
            host = host,
            port = port,
            serverNoisePubkey = noisePubkey,
            metadata = metadata
        )
    }
    
    fun discoverEndpoint(pubkey: String): MockEndpoint? {
        return endpoints[pubkey]
    }
    
    fun removeEndpoint(pubkey: String) {
        endpoints.remove(pubkey)
    }
    
    fun clear() {
        endpoints.clear()
    }
}

// MARK: - Test Data Generators

/**
 * Generates test data for E2E tests
 */
object TestDataGenerator {
    
    /**
     * Generate a valid test payment request
     */
    fun createPaymentRequest(
        from: String = "test_payer",
        to: String = "test_payee",
        amount: Long = 1000
    ): Pair<String, Map<String, Any>> {
        val receiptId = "rcpt_test_${UUID.randomUUID()}"
        val request = mapOf(
            "receipt_id" to receiptId,
            "payer_pubkey" to from,
            "payee_pubkey" to to,
            "method_id" to "lightning",
            "amount_sats" to amount,
            "created_at" to java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(Date())
        )
        return receiptId to request
    }
    
    /**
     * Generate a valid test receipt confirmation
     */
    fun createReceiptConfirmation(
        receiptId: String,
        payee: String = "test_payee"
    ): Map<String, Any> {
        return mapOf(
            "receipt_id" to receiptId,
            "confirmed_at" to java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(Date()),
            "payee_pubkey" to payee,
            "status" to "confirmed"
        )
    }
    
    /**
     * Generate random bytes for test keys
     */
    fun randomBytes(count: Int): ByteArray {
        return ByteArray(count) { (0..255).random().toByte() }
    }
    
    /**
     * Generate a mock Noise public key (hex string)
     */
    fun mockNoisePubkey(): String {
        return randomBytes(32).joinToString("") { "%02x".format(it) }
    }
}

// MARK: - Test Errors

/**
 * Test-specific exceptions
 */
sealed class TestError(message: String) : Exception(message) {
    object Timeout : TestError("Test operation timed out")
    class SetupFailed(msg: String) : TestError("Test setup failed: $msg")
    class AssertionFailed(msg: String) : TestError("Assertion failed: $msg")
    class MockError(msg: String) : TestError("Mock error: $msg")
}

// MARK: - Network Test Helpers

/**
 * Helpers for network-based testing
 */
object NetworkTestHelpers {
    
    /**
     * Check if a port is available
     */
    fun isPortAvailable(port: Int): Boolean {
        return try {
            java.net.ServerSocket(port).use { true }
        } catch (e: Exception) {
            false
        }
    }
    
    /**
     * Find an available port
     */
    fun findAvailablePort(): Int {
        return TestConfig.randomPort()
    }
    
    /**
     * Create a local loopback address for testing
     */
    fun loopbackAddress(port: Int): String {
        return "127.0.0.1:$port"
    }
}

// MARK: - Assertion Helpers

/**
 * Custom assertions for payment tests
 */
object PaymentAssertions {
    
    /**
     * Assert that a receipt was created correctly
     */
    fun assertReceiptValid(
        receipt: MockReceiptStore.MockReceipt,
        expectedPayer: String,
        expectedPayee: String
    ) {
        assert(receipt.id.isNotEmpty()) { "Receipt ID should not be empty" }
        assert(receipt.payerPubkey == expectedPayer) { "Payer should match" }
        assert(receipt.payeePubkey == expectedPayee) { "Payee should match" }
        assert(receipt.amountSats > 0) { "Amount should be positive" }
    }
    
    /**
     * Assert that an endpoint is valid
     */
    fun assertEndpointValid(
        endpoint: MockDirectoryService.MockEndpoint,
        expectedHost: String? = null,
        expectedPort: Int? = null
    ) {
        assert(endpoint.recipientPubkey.isNotEmpty()) { "Pubkey should not be empty" }
        assert(endpoint.host.isNotEmpty()) { "Host should not be empty" }
        assert(endpoint.port > 0) { "Port should be positive" }
        assert(endpoint.serverNoisePubkey.isNotEmpty()) { "Noise pubkey should not be empty" }
        
        expectedHost?.let { assert(endpoint.host == it) }
        expectedPort?.let { assert(endpoint.port == it) }
    }
}

// MARK: - Mock Server

/**
 * Mock server configuration
 */
data class MockServerConfig(
    val port: Int,
    val noisePubkey: String,
    val maxConnections: Int = 100,
    val timeout: Long = 60000L
)

/**
 * Mock Noise server for testing
 */
class MockNoiseServer {
    var isRunning = false
        private set
    var activeConnections = 0
        private set
    private val connections = mutableSetOf<String>()
    var maxConnections = 100
    var receiptCallback: ((Map<String, Any>) -> String)? = null
    
    fun start(port: Int): Boolean {
        isRunning = true
        return true
    }
    
    fun stop() {
        isRunning = false
        connections.clear()
        activeConnections = 0
    }
    
    fun acceptConnection(): String? {
        if (!isRunning || activeConnections >= maxConnections) return null
        val clientId = UUID.randomUUID().toString()
        connections.add(clientId)
        activeConnections++
        return clientId
    }
    
    fun disconnectClient(clientId: String) {
        if (connections.remove(clientId)) {
            activeConnections--
        }
    }
    
    fun simulateClientDisconnect(clientId: String) {
        disconnectClient(clientId)
    }
    
    fun processMessage(clientId: String, messageType: String, payload: Map<String, Any>): Map<String, Any>? {
        if (!connections.contains(clientId)) return null
        
        return when (messageType) {
            "payment_request" -> {
                val callback = receiptCallback
                if (callback != null) {
                    val receiptId = callback(payload)
                    mapOf("status" to "confirmed", "receipt_id" to receiptId)
                } else {
                    mapOf(
                        "status" to "confirmed",
                        "receipt_id" to (payload["receipt_id"] ?: UUID.randomUUID().toString())
                    )
                }
            }
            else -> mapOf("status" to "error", "error" to "unknown_message_type")
        }
    }
}

File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemoTests/TestHelpers.swift
// TestHelpers.swift
// Test Helpers for PaykitDemo E2E Tests
//
// Provides common utilities, mock services, and test fixtures for E2E testing.

import XCTest
import Foundation

// MARK: - Test Configuration

/// Configuration for E2E tests
struct TestConfig {
    /// Test timeout for async operations
    static let defaultTimeout: TimeInterval = 30.0
    
    /// Short timeout for quick operations
    static let shortTimeout: TimeInterval = 5.0
    
    /// Port range for test servers
    static let testPortRange = 10000...60000
    
    /// Test user identifiers
    static let testUserAlice = "alice_test_\(UUID().uuidString.prefix(8))"
    static let testUserBob = "bob_test_\(UUID().uuidString.prefix(8))"
    
    /// Generate a random available port
    static func randomPort() -> UInt16 {
        UInt16.random(in: 10000...60000)
    }
}

// MARK: - Mock Key Manager

/// Mock key manager for testing
class MockKeyManager {
    private var identities: [String: MockIdentity] = [:]
    private var currentIdentity: String?
    
    struct MockIdentity {
        let nickname: String
        let publicKeyZ32: String
        let secretKey: Data
        
        init(nickname: String) {
            self.nickname = nickname
            self.publicKeyZ32 = "z32_\(UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(52))"
            self.secretKey = Data((0..<32).map { _ in UInt8.random(in: 0...255) })
        }
    }
    
    func createIdentity(nickname: String) -> MockIdentity {
        let identity = MockIdentity(nickname: nickname)
        identities[nickname] = identity
        return identity
    }
    
    func setCurrentIdentity(_ nickname: String) -> Bool {
        guard identities[nickname] != nil else { return false }
        currentIdentity = nickname
        return true
    }
    
    func getCurrentIdentity() -> MockIdentity? {
        guard let current = currentIdentity else { return nil }
        return identities[current]
    }
    
    func getPublicKey() -> String? {
        return getCurrentIdentity()?.publicKeyZ32
    }
}

// MARK: - Mock Receipt Store

/// Mock receipt store for testing
class MockReceiptStore {
    private var receipts: [String: MockReceipt] = [:]
    
    struct MockReceipt {
        let id: String
        let payerPubkey: String
        let payeePubkey: String
        let amountSats: UInt64
        let createdAt: Date
        let status: String
    }
    
    func storeReceipt(_ receipt: MockReceipt) {
        receipts[receipt.id] = receipt
    }
    
    func getReceipt(_ id: String) -> MockReceipt? {
        return receipts[id]
    }
    
    func getAllReceipts() -> [MockReceipt] {
        return Array(receipts.values)
    }
    
    func clear() {
        receipts.removeAll()
    }
}

// MARK: - Mock Directory Service

/// Mock directory service for testing endpoint discovery
class MockDirectoryService {
    private var endpoints: [String: MockEndpoint] = [:]
    
    struct MockEndpoint {
        let recipientPubkey: String
        let host: String
        let port: UInt16
        let serverNoisePubkey: String
        let metadata: String?
    }
    
    func publishEndpoint(
        pubkey: String,
        host: String,
        port: UInt16,
        noisePubkey: String,
        metadata: String? = nil
    ) {
        endpoints[pubkey] = MockEndpoint(
            recipientPubkey: pubkey,
            host: host,
            port: port,
            serverNoisePubkey: noisePubkey,
            metadata: metadata
        )
    }
    
    func discoverEndpoint(pubkey: String) -> MockEndpoint? {
        return endpoints[pubkey]
    }
    
    func removeEndpoint(pubkey: String) {
        endpoints.removeValue(forKey: pubkey)
    }
    
    func clear() {
        endpoints.removeAll()
    }
}

// MARK: - Test Data Generators

/// Generates test data for E2E tests
struct TestDataGenerator {
    /// Generate a valid test payment request
    static func createPaymentRequest(
        from payer: String = "test_payer",
        to payee: String = "test_payee",
        amount: UInt64 = 1000
    ) -> (receiptId: String, request: [String: Any]) {
        let receiptId = "rcpt_test_\(UUID().uuidString)"
        let request: [String: Any] = [
            "receipt_id": receiptId,
            "payer_pubkey": payer,
            "payee_pubkey": payee,
            "method_id": "lightning",
            "amount_sats": amount,
            "created_at": ISO8601DateFormatter().string(from: Date())
        ]
        return (receiptId, request)
    }
    
    /// Generate a valid test receipt confirmation
    static func createReceiptConfirmation(
        for receiptId: String,
        payee: String = "test_payee"
    ) -> [String: Any] {
        return [
            "receipt_id": receiptId,
            "confirmed_at": ISO8601DateFormatter().string(from: Date()),
            "payee_pubkey": payee,
            "status": "confirmed"
        ]
    }
    
    /// Generate random bytes for test keys
    static func randomBytes(_ count: Int) -> Data {
        Data((0..<count).map { _ in UInt8.random(in: 0...255) })
    }
    
    /// Generate a mock Noise public key (hex string)
    static func mockNoisePubkey() -> String {
        randomBytes(32).map { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - Async Test Helpers

/// Helpers for async testing
extension XCTestCase {
    /// Wait for an async operation with timeout
    func waitForAsync<T>(
        timeout: TimeInterval = TestConfig.defaultTimeout,
        operation: @escaping () async throws -> T
    ) throws -> T {
        let expectation = self.expectation(description: "Async operation")
        var result: Result<T, Error>?
        
        Task {
            do {
                let value = try await operation()
                result = .success(value)
            } catch {
                result = .failure(error)
            }
            expectation.fulfill()
        }
        
        wait(for: [expectation], timeout: timeout)
        
        switch result {
        case .success(let value):
            return value
        case .failure(let error):
            throw error
        case .none:
            throw TestError.timeout
        }
    }
    
    /// Assert that an async operation completes without error
    func assertNoThrowAsync(
        timeout: TimeInterval = TestConfig.defaultTimeout,
        _ operation: @escaping () async throws -> Void,
        file: StaticString = #file,
        line: UInt = #line
    ) {
        let expectation = self.expectation(description: "Async operation")
        var caughtError: Error?
        
        Task {
            do {
                try await operation()
            } catch {
                caughtError = error
            }
            expectation.fulfill()
        }
        
        wait(for: [expectation], timeout: timeout)
        
        if let error = caughtError {
            XCTFail("Unexpected error: \(error)", file: file, line: line)
        }
    }
}

// MARK: - Test Errors

enum TestError: Error, LocalizedError {
    case timeout
    case setupFailed(String)
    case assertionFailed(String)
    case mockError(String)
    
    var errorDescription: String? {
        switch self {
        case .timeout:
            return "Test operation timed out"
        case .setupFailed(let msg):
            return "Test setup failed: \(msg)"
        case .assertionFailed(let msg):
            return "Assertion failed: \(msg)"
        case .mockError(let msg):
            return "Mock error: \(msg)"
        }
    }
}

// MARK: - Network Test Helpers

/// Helpers for network-based testing
struct NetworkTestHelpers {
    /// Check if a port is available
    static func isPortAvailable(_ port: UInt16) -> Bool {
        // Simple check - try to create a socket
        // In real tests, this would use proper socket APIs
        return true
    }
    
    /// Find an available port
    static func findAvailablePort() -> UInt16 {
        // Return a random port from the test range
        // In production, this would actually check availability
        return TestConfig.randomPort()
    }
    
    /// Create a local loopback address for testing
    static func loopbackAddress(port: UInt16) -> String {
        "127.0.0.1:\(port)"
    }
}

// MARK: - Assertion Helpers

/// Custom assertions for payment tests
struct PaymentAssertions {
    /// Assert that a receipt was created correctly
    static func assertReceiptValid(
        _ receipt: MockReceiptStore.MockReceipt,
        expectedPayer: String,
        expectedPayee: String,
        file: StaticString = #file,
        line: UInt = #line
    ) {
        XCTAssertFalse(receipt.id.isEmpty, "Receipt ID should not be empty", file: file, line: line)
        XCTAssertEqual(receipt.payerPubkey, expectedPayer, "Payer should match", file: file, line: line)
        XCTAssertEqual(receipt.payeePubkey, expectedPayee, "Payee should match", file: file, line: line)
        XCTAssertGreaterThan(receipt.amountSats, 0, "Amount should be positive", file: file, line: line)
    }
    
    /// Assert that an endpoint is valid
    static func assertEndpointValid(
        _ endpoint: MockDirectoryService.MockEndpoint,
        expectedHost: String? = nil,
        expectedPort: UInt16? = nil,
        file: StaticString = #file,
        line: UInt = #line
    ) {
        XCTAssertFalse(endpoint.recipientPubkey.isEmpty, "Pubkey should not be empty", file: file, line: line)
        XCTAssertFalse(endpoint.host.isEmpty, "Host should not be empty", file: file, line: line)
        XCTAssertGreaterThan(endpoint.port, 0, "Port should be positive", file: file, line: line)
        XCTAssertFalse(endpoint.serverNoisePubkey.isEmpty, "Noise pubkey should not be empty", file: file, line: line)
        
        if let expectedHost = expectedHost {
            XCTAssertEqual(endpoint.host, expectedHost, file: file, line: line)
        }
        if let expectedPort = expectedPort {
            XCTAssertEqual(endpoint.port, expectedPort, file: file, line: line)
        }
    }
}
